### PR TITLE
fixed `get_parameter_list` - account for masking instructions

### PR DESCRIPTION
**Pull request by Impelon**
This is my attempt at providing a fix for #49.

Unfortunately it has two flaws, meaning it cannot really be merged, but hopefully it can serve as a starting point for a better solution. The two flaws being:

1. The comparatively minor issue is that it does not always behave correctly for log messages that do not match the provided template. It may even throw a `StopIteration` exception. But I don't think anyone expects meaningful results in that case anyways. Nevertheless, here is an example:
```python
>>> from drain3 import TemplateMiner, masking
>>> parser = TemplateMiner()
>>> parser.config.masking_instructions = [masking.MaskingInstruction(r"\d+\.\d+", "float")]
>>> parser.get_parameter_list("<float> <float>", "0.15 0.1 test 0.1 2.3")
['0.15', '0.1', '0.1', '2.3']
>>> parser.get_parameter_list("<*><float>", "")
Traceback (most recent call last):
[...]
StopIteration
```
2. I only now realized that this fix makes the implicit assumption that `<*>` or rather a mask of `*` (with any prefix/suffix) is always generated by Drain itsself and not by a `MaskingInstruction`. That assumption is obviously flawed.

The solution may also be more complicated that it needs to be. I'll leave some comments on the idea behind the approach.
Sadly I'm not sure how to overcome the 2. flaw, but I hope this can help in finding a better solution.
Feel free to change anything!

### CODE REVIEW

File: drain3/template_miner.py [hunk 1]
```diff
@@ -173,21 +173,30 @@ def match(self, log_message: str, full_search_strategy="never") -> LogCluster:
     def get_parameter_list(self, log_template: str, content: str):
         escaped_prefix = re.escape(self.config.mask_prefix)
         escaped_suffix = re.escape(self.config.mask_suffix)
-        template_regex = re.sub(escaped_prefix + r".+?" + escaped_suffix, self.drain.param_str, log_template)
-        if self.drain.param_str not in template_regex:
-            return []
-        template_regex = re.escape(template_regex)
-        template_regex = re.sub(r'\\ +', r'\\s+', template_regex)
-        template_regex = "^" + template_regex.replace(escaped_prefix + r"\*" + escaped_suffix, "(.*?)") + "$"
+        content_masked =  re.sub(escaped_prefix + r".+?" + escaped_suffix, lambda match: self.config.mask_prefix + "*" + self.config.mask_suffix, content)
```
Comment (Impelon) [inline drain3/template_miner.py:176]: It is assumed that it is safe to place masks inside the log message without changing the behavior of masking-patterns.

As masks are later used to store parameters, the first step is to replace the content of any existing masks in the message.

File: drain3/template_miner.py [hunk 2]
```diff
@@ -173,21 +173,28 @@ def match(self, log_message: str, full_search_strategy="never") -> LogCluster:
...
+        for instruction in self.config.masking_instructions:
+            content_masked = instruction.regex.sub(lambda match: self.config.mask_prefix + "(" + re.escape(match.group(0)) + ")" + self.config.mask_suffix, content_masked)
```
Comment (Impelon) [inline drain3/template_miner.py:178]: Replace occurrences of a parameter identified by a masking instruction with a mask containing the parameter.

The parameter is escaped additionally in case it contains `)` as a delimiter.

File: drain3/template_miner.py [hunk 3]
```diff
@@ -173,21 +173,28 @@ def match(self, log_message: str, full_search_strategy="never") -> LogCluster:
...
+        masking_parameter_list = [re.sub(r"\\(.)", r"\1", match.group(1)) for match in re.finditer(escaped_prefix + r"\((.+?)(?<!\\)\)" + escaped_suffix, content_masked)]
```
Comment (Impelon) [inline drain3/template_miner.py:179]: The parameters are extracted (in the correct order!) from all masks containing parameters (which can be easily identified as the only masks like `<(...)>`).

The escaping of the parameter is undone now.

File: drain3/template_miner.py [hunk 4]
```diff
@@ -173,21 +173,28 @@ def match(self, log_message: str, full_search_strategy="never") -> LogCluster:
...
+
+        escaped_param_str = re.escape(self.drain.param_str)
+        template_regex = re.escape(log_template)
+        if escaped_param_str not in template_regex:
+            return masking_parameter_list
+        masking_parameter_iterator = iter(masking_parameter_list)
+        template_regex = re.sub(re.escape(escaped_prefix) + r"(?!\\[*]).+?" + re.escape(escaped_suffix), lambda x: "(" + next(masking_parameter_iterator) + ")", template_regex)
```
Comment (Impelon) [inline drain3/template_miner.py:186]: Only replace occurrences of masks introduced by `MaskingInstruction` with the parameters in order.
(Note: This is where the faulty assumption is made that all masks introduced by `MaskingInstruction` do not have `*` as the value of `mask_with`.)

**Comment by davidohana** [comment]
Please see my comments regarding this PR in issue #49

File: drain3/template_miner.py [hunk 5]
```diff
@@ -170,24 +172,46 @@ def match(self, log_message: str, full_search_strategy="never") -> LogCluster:
...
+            # This way they cannot be confused with any real parameters.
+            masking_parameters[match.group(1)] = None
+        for instruction in self.config.masking_instructions:
+            def mask_parameter(match):
+                id = instruction.mask_with
+                while id in masking_parameters.keys():
+                    id = valid_mask_generator(id)
+                masking_parameters[id] = match.group(0)
+                return self.config.mask_prefix + id + self.config.mask_suffix
+            content_masked = instruction.regex.sub(mask_parameter, content_masked)
```
Comment (Impelon) [inline drain3/template_miner.py:194]: Parameters are masked with a unique mask, that can later be looked up to get the actual parameter.

To make sure other masking-patterns do not interfere with the added mask, an optional function `valid_mask_generator` can be provided, that should generate a new valid mask (and is provided with the mask of an unsuccessful attempt to create an unique mask).
The default for `valid_mask_generator` should work in most cases, as it just adds another random character from the existing mask.

**Comment by Impelon** [comment]
I've included the tests from #51 in this change and added some relevant `MaskingInstruction`-objects, so the new method can actually do its work.

There were some problems with the previous iteration, where it failed to extract the correct parameters, because the masking-patterns I used myself _were_ interfering with the temporary masks added. 
I've circumvented that problem by using a unique temporary mask as an identifier to get the actual parameter, instead of storing the parameter in the temporary mask. As an added benefit, no messy escaping of masks is needed any more.
TL;DR: The parameter extraction actually works correctly now.

Sadly this still suffers from the two flaws laid out above.

**Comment by Impelon** [comment]
Actually your comment in #49 gave me an idea:
> For the first approach, the issue of non-unique mapping from mask name e.g. `<NUM>` to a regex can be resolved using an or (`|`) operator in the mask matching regex, and the **user may use unique mask names if he/she wants to avoid that.**

The final change made here is simple; if the user wants to use the new method to extract parameters identified by `MaskingInstruction`, they need to provide it with a `mask_value` other than `*`.
Otherwise the parameter identified by that `MaskingInstruction` will just be handled as before. 

Therefore this circumvents the 2. flaw and should be a strict improvement (accuracy-wise) for valid templates.
As my previous attempts at solving this have had some problems, I think more tests are should be added as well. 

I've added some inline-comments and a docstring to document the new method; please feel free to modify!

**Comment by Impelon** [comment]
I've reworked the tests a bit, added just a few more tests, made some fixes according to those tests.
As a comparison, here is an overview of which tests the different versions of `get_parameter_list` fail:

### v0.9.8:
- Template: `<hdfs_uri>:<integer>+<integer>`
  Message: `hdfs://msra-sa-41:9000/pageinput2.txt:671088640+134217728`
- Template: `<float>.<*>`
  Message: `0.15.Test`
- Template: `<float>.<*>.<float>`
  Message: `0.15.Test.0.2`
- Template: `<*words*> <*>`
  Message: `multiple word pattern <*words*>`
- Template: `<*> <*>`
  Message: `HelloWorld <anything>`
- Template: `<float> <float>` (invalid template for message)
  Message: `0.15 10.16 3.19`
### #51
- Template: `<float>.<*>.<float>`
  Message: `0.15.Test.0.2`
- Template: `<*words*> <*>`
  Message: `multiple word pattern <*words*>`
- Template: `<*> <*>`
  Message: `HelloWorld <anything>`
### #50
- Invalid templates produce invalid results and may raise `StopIteration`-exceptions:
  - Template: `<float> <float>`
    Message: `0.15 10.16 3.19`
  - Template: `<float> <float>`
    Message: `0.15 10.16 test 3.19`

**Comment by Impelon** [comment]
I tried using this in my application and soon found out more problems that need to be tested (that also produce problems with the version in v0.9.8).

Since this has become quite a mess and in the process of fixing the problems I've found I've moved towards the solution proposed by you in #49, I will close this and open another PR when done instead.
